semantics/src/main/kotlin/com/strumenta/kolasu/semantics/scope/description/ScopeDescription.kt

```diff
@@ -0,0 +1,134 @@
+package com.strumenta.kolasu.semantics.scope.description
+
+import com.strumenta.kolasu.model.Node
+import com.strumenta.kolasu.model.PossiblyNamed
+import com.strumenta.kolasu.model.ReferenceByName
+import com.strumenta.kolasu.semantics.symbol.description.SymbolDescription
+
+/**
+ * Utility function for defining scope descriptions.
+ **/
+fun scope(
+    ignoreCase: Boolean = false,
+    init: ScopeDescriptionApi.() -> Unit
+): ScopeDescription = ScopeDescription(ignoreCase).apply(init)
+
+/**
+ * A scope description containing all relevant information
+ * concerning the visible elements at a given point in space or time.
+ * - `ignoreCase`: whether the names associated to the symbols
+ * in this scope should be handled as case-sensitive or not;
+ * - `nodes`: the local symbols contained in this scope and
+ * their associated name;
+ * - `identifiers`: the external symbols contained in this scope
+ * and their associated name;
+ * - `parent`: the scope description *preceding* the current one - used
+ * as delegate if no symbol can be found for a given name in the current scope;
+ *
+ * Each scope description can be configured through the methods defined in
+ * the `ScopeDescriptionApi`. References can be resolved invoking `resolve(reference)`.
+ *
+ * Given a reference, the scope provider will follow the following plan:
+ * - try to retrieve a local symbol associated with the given reference name;
+ * - if not found, try to retrieve a global symbol associated with the given reference name;
+ * - if not found, delegate the resolution to its parent (if any)
+ *
+ **/
+class ScopeDescription(
+    private val ignoreCase: Boolean = false
+) : ScopeDescriptionApi {
+    private var parent: ScopeDescription? = null
+    private val identifiers: MutableMap<String, String> = mutableMapOf()
```

Comment:
Is identifiers a map from name to identifiers? If so I suggest renaming it `namedToIdentifiers`, because it would be easy to invert name and identifier when using it

Reply:
I agree, will do it